### PR TITLE
1110 1113 1119 836 - Fix and improve api for contextmenu with datagrid

### LINKED ISSUE #836: IdsPopup: Arrow not aligned the first time
**Describe the bug**
When run first time the popup arrow not align. It has wrong position calculated in `placeArrow()` method.

**To Reproduce**
- Open with any editor (ie: vs-code) `src/components/ids-pie-chart/ids-pie-chart.ts`
- Search for commented line `// TODO: Find a way to work without initial visible call`
- Make disable/comment next line `tooltip.popup.visible = true;`
- build/run and goto: http://localhost:4300/ids-pie-chart/example.html
- Hover on any pie slice to show tooltip
- See the bug now, very first time the tooltip arrow not align

**Expected behavior**
Tooltip/popup arrow should not misaligned first or any time

Related #736

### LINKED ISSUE #1110: Popupmenu: Improve Dynamic Menu API
**Describe the bug**
When trying to use a dynamic menu in the data grid. The API could be improved and also the position changes

**To Reproduce**

1. Replace the attached example on 
2. Go to http://localhost:4300/ids-data-grid/contextmenu-thru-data.html
3. Right click the grid - see position error

**Expected behavior**
1. on http://localhost:4300/ids-data-grid/contextmenu-thru-data.html (without the update). Remove the arrow from the header and body popupmenu as it doesnt point at anything usefully
2. when using the new dynamic example the position is "centered" - should work the same and be at the mouse/cell position when dynamic or not
3. Instead of setting `const menuEl = e.detail?.data?.menuEl; menuEl.data = headerMenuData;` it should work to set `dataGrid.menuData = newData;` or dataGrid.headerMenuData = newData;`

**Version**
- beta 4

**Screenshots**
Screenshot 2023-01-27 at 11 40 06 AM

**Example**

```js
import '../ids-data-grid';
import booksJSON from '../../../assets/data/books.json';
import menuContentsJSON from '../../../assets/data/menu-contents.json';

// Example for populating the DataGrid
const dataGrid: any = document.querySelector('#data-grid-contextmenu-thru-data');
const container: any = document.querySelector('ids-container');

if (dataGrid) {
  (async function init() {
    // Set Locale and wait for it to load
    await container?.setLocale('en-US');

    // Do an ajax request
    const url: any = booksJSON;
    const menuUrl: any = menuContentsJSON;

    // Header contextmenu data
    const headerMenuData = {
      id: 'grid-header-menu',
      contents: [{
        id: 'actions-group',
        items: [
          { id: 'actions-split', value: 'actions-split', text: 'Split' },
          { id: 'actions-sort', value: 'actions-sort', text: 'Sort' },
          { id: 'actions-hide', value: 'actions-hide', text: 'Hide' }
        ]
      }],
    };

    // Set up columns
    const columns = [];
    columns.push({
      id: 'selectionCheckbox',
      name: 'selection',
      sortable: false,
      resizable: false,
      formatter: dataGrid.formatters.selectionCheckbox,
      align: 'center'
    });
    columns.push({
      id: 'rowNumber',
      name: '#',
      formatter: dataGrid.formatters.rowNumber,
      sortable: false,
      readonly: true,
      width: 65
    });
    columns.push({
      id: 'description',
      name: 'Description',
      field: 'description',
      sortable: true,
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'ledger',
      name: 'Ledger',
      field: 'ledger',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'publishDate',
      name: 'Pub. Date',
      field: 'publishDate',
      formatter: dataGrid.formatters.date
    });
    columns.push({
      id: 'publishTime',
      name: 'Pub. Time',
      field: 'publishDate',
      formatter: dataGrid.formatters.time
    });
    columns.push({
      id: 'price',
      name: 'Price',
      field: 'price',
      formatter: dataGrid.formatters.decimal,
      formatOptions: { locale: 'en-US' } // Data Values are in en-US
    });
    columns.push({
      id: 'bookCurrency',
      name: 'Currency',
      field: 'bookCurrency',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'transactionCurrency',
      name: 'Transaction Currency',
      field: 'transactionCurrency',
      formatter: dataGrid.formatters.text,
    });
    columns.push({
      id: 'integer',
      name: 'Price (Int)',
      field: 'price',
      formatter: dataGrid.formatters.integer,
      formatOptions: { locale: 'en-US' } // Data Values are in en-US
    });
    columns.push({
      id: 'location',
      name: 'Location',
      field: 'location',
      formatter: dataGrid.formatters.hyperlink,
      href: '#'
    });
    columns.push({
      id: 'postHistory',
      name: 'Post History',
      field: 'postHistory',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'active',
      name: 'Active',
      field: 'active',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'convention',
      name: 'Convention',
      field: 'convention',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'methodSwitch',
      name: 'Method Switch',
      field: 'methodSwitch',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'trackDeprecationHistory',
      name: 'Track Deprecation History',
      field: 'trackDeprecationHistory',
      formatter: dataGrid.formatters.dropdown
    });
    columns.push({
      id: 'useForEmployee',
      name: 'Use For Employee',
      field: 'useForEmployee',
      formatter: dataGrid.formatters.password
    });
    columns.push({
      id: 'deprecationHistory',
      name: 'Deprecation History',
      field: 'deprecationHistory',
      formatter: dataGrid.formatters.text
    });

    dataGrid.columns = columns;
    const setData = async () => {
      let counter = 0;
      const res = await fetch(url);
      const menuRes = await fetch(menuUrl);

      const data = await res.json();
      const menuData = await menuRes.json();

      dataGrid.data = data;
      dataGrid.menuData = menuData;
      dataGrid.headerMenuData = headerMenuData;

      dataGrid.addEventListener('beforemenushow', (e: any) => {
        console.info('before contextmenu show', e.detail);
        counter++;

        const headerMenuData2 = {
          id: 'grid-header-menu',
          contents: [{
            id: 'actions-group',
            items: [
              { id: 'actions-split', value: 'actions-split', text: 'Split2' },
              { id: 'actions-sort', value: 'actions-sort', text: 'Sort2' },
              { id: 'actions-hide', value: 'actions-hide', text: 'Hide2' }
            ]
          }],
        };
        const menuEl = e.detail?.data?.menuEl;
        if (menuEl) menuEl.data = headerMenuData2;
      });

      dataGrid.addEventListener('menushow', (e: any) => {
        console.info('After contextmenu show', e.detail);
      });

      dataGrid.addEventListener('menuselected', (e: any) => {
        console.info('contextmenu item selected', e.detail);
      });
    };

    setData();
  }());
}
```

**Additional context**
M3 Email

### LINKED ISSUE #1113: Popupmenu: Improve Event Support
**Describe the bug**
The current api for popupmenus in datagrid has a bug where the events will stack up. Will also like to be able to distinguish right clicking between:  header vs header filter, and normal cell vs editable element

**To Reproduce**

1. Copy attached example over the existing one and run
1. Right click on the header and select a menu
1. Right click on the header filter and select a menu
1. Right click on a cell and select a menu
1. Right click on different cell and select a menu

In addition add the following to the m3-example.html
```
      dataGrid.addEventListener('beforemenushow', (e: any) => {
        console.info(detail.data.type);
      })
```
1. try to right click the header -> can see the type this is working
1. try to right click the header filter input -> need a new type for this
1.  try to right click a cell -> can see the type this is working
1.  try to right click an inline editable cell -> need a new type for this
1. Add md Docs for this and tests

**Expected behavior**
New types in events and events not stacking up

**Additional context**
M3 - email..

Code:
```js
import '../ids-data-grid';
import booksJSON from '../../../assets/data/books.json';
import menuContentsJSON from '../../../assets/data/menu-contents.json';

// Example for populating the DataGrid
const dataGrid: any = document.querySelector('#data-grid-contextmenu-thru-data');
const container: any = document.querySelector('ids-container');

if (dataGrid) {
  (async function init() {
    // Set Locale and wait for it to load
    await container?.setLocale('en-US');

    // Do an ajax request
    const url: any = booksJSON;
    const menuUrl: any = menuContentsJSON;

    // Header contextmenu data
    const headerMenuData = {
      id: 'grid-header-menu',
      contents: [{
        id: 'actions-group',
        items: [
          { id: 'actions-split', value: 'actions-split', text: 'Split' },
          { id: 'actions-sort', value: 'actions-sort', text: 'Sort' },
          { id: 'actions-hide', value: 'actions-hide', text: 'Hide' }
        ]
      }],
    };

    // Set up columns
    const columns = [];
    columns.push({
      id: 'selectionCheckbox',
      name: 'selection',
      sortable: false,
      resizable: false,
      formatter: dataGrid.formatters.selectionCheckbox,
      align: 'center'
    });
    columns.push({
      id: 'rowNumber',
      name: '#',
      formatter: dataGrid.formatters.rowNumber,
      sortable: false,
      readonly: true,
      width: 65
    });
    columns.push({
      id: 'description',
      name: 'Description',
      field: 'description',
      sortable: true,
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'ledger',
      name: 'Ledger',
      field: 'ledger',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'publishDate',
      name: 'Pub. Date',
      field: 'publishDate',
      formatter: dataGrid.formatters.date
    });
    columns.push({
      id: 'publishTime',
      name: 'Pub. Time',
      field: 'publishDate',
      formatter: dataGrid.formatters.time
    });
    columns.push({
      id: 'price',
      name: 'Price',
      field: 'price',
      formatter: dataGrid.formatters.decimal,
      formatOptions: { locale: 'en-US' } // Data Values are in en-US
    });
    columns.push({
      id: 'bookCurrency',
      name: 'Currency',
      field: 'bookCurrency',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'transactionCurrency',
      name: 'Transaction Currency',
      field: 'transactionCurrency',
      formatter: dataGrid.formatters.text,
    });
    columns.push({
      id: 'integer',
      name: 'Price (Int)',
      field: 'price',
      formatter: dataGrid.formatters.integer,
      formatOptions: { locale: 'en-US' } // Data Values are in en-US
    });
    columns.push({
      id: 'location',
      name: 'Location',
      field: 'location',
      formatter: dataGrid.formatters.hyperlink,
      href: '#'
    });
    columns.push({
      id: 'postHistory',
      name: 'Post History',
      field: 'postHistory',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'active',
      name: 'Active',
      field: 'active',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'convention',
      name: 'Convention',
      field: 'convention',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'methodSwitch',
      name: 'Method Switch',
      field: 'methodSwitch',
      formatter: dataGrid.formatters.text
    });
    columns.push({
      id: 'trackDeprecationHistory',
      name: 'Track Deprecation History',
      field: 'trackDeprecationHistory',
      formatter: dataGrid.formatters.dropdown
    });
    columns.push({
      id: 'useForEmployee',
      name: 'Use For Employee',
      field: 'useForEmployee',
      formatter: dataGrid.formatters.password
    });
    columns.push({
      id: 'deprecationHistory',
      name: 'Deprecation History',
      field: 'deprecationHistory',
      formatter: dataGrid.formatters.text
    });

    dataGrid.columns = columns;
    const setData = async () => {
      let counter = 0;
      const res = await fetch(url);
      const menuRes = await fetch(menuUrl);

      const data = await res.json();
      const menuData = await menuRes.json();

      dataGrid.data = data;
      dataGrid.menuData = menuData;
      dataGrid.headerMenuData = headerMenuData;

      dataGrid.addEventListener('beforemenushow', (e: any) => {
        console.info('before contextmenu show', e.detail);
        counter++;
        if (counter % 2 === 0) {
          dataGrid.menuData = headerMenuData;
        } else {
          dataGrid.menuData = menuData;
        }
        // e.detail.response(false);
      });

      dataGrid.addEventListener('menushow', (e: any) => {
        console.info('After contextmenu show', e.detail);
      });

      dataGrid.addEventListener('menuselected', (e: any) => {
        console.info('contextmenu item selected', e.detail);
      });
    };

    setData();
  }());
}

```

### LINKED ISSUE #1119: IdsDatagrid: Context menu cannot  display all items
**Describe the bug**
When the context menu has a lot of items, not all items are display or inaccessible.

**To Reproduce**

1. Open enterprise-wc
1. Update contextmenu-thru-data.ts
2. Add more to items in id: 'actions-group'
3. npm start
4. Open contextmenu-thru-data.html
5. Right click on the header
6. Error, context menu items is cut

**Expected behavior**
Contextmenu should have a scroll bar when there are a lot of items

**Version**

ids enterprise wc 1.0.0-beta.3

**Screenshots**
image

**Platform**
Infor Application/Team Name: M3 UX
OS Version: Windows 11
Browser Name: [chrome
Browser Version: 109.0.5414.75 (Official Build) (64-bit)

**Additional context**

### PR DESCRIPTION
**Pull request by deep7102**
**Explain the details for making this change. What existing problem does the pull request solve?**
Fixed and improved api for contextmenu with datagrid.

**Related github/jira issue (required)**:
Closes #1110
Closes #1113
Closes #1119
Closes #836

**Steps necessary to review your pull request (required)**:
Pull this branch and build/run the demo app

Navigate to: http://localhost:4300/ids-data-grid/contextmenu-thru-data.html
- Right click the grid (header and body cells)
- See there should not be arrow with the contextmenu
- Right click each area to grid
- See in console log for it's type (for example `body-cell`, `header-title`, `header-filter`, `header-filter-button`)
- Click on any cell in column `Description` to get in edit mode
- While in edit mode right click and see console log for it's type `body-cell-editor`

Navigate to: http://localhost:4300/ids-data-grid/contextmenu-before-show.html
- Right click on grid body cells
- It loads dynamic menu content
- See demo the api use as `dataGrid.menuData = newData;`
- Also see contextmenu position should be left top alignment

Navigate to: http://localhost:4300/ids-data-grid/contextmenu-overflow.html
- Right click on grid
- See should not overflow to display all items with contextmenu

Navigate to: http://localhost:4300/ids-pie-chart/example.html
- Hover on any pie slice to show tooltip
- See tooltip arrow for very first time it should not misalign

**Included in this Pull Request**:
- [x] A note to the change log.

### CODE REVIEW

File: src/components/ids-popup/ids-popup.ts [hunk 1]
```diff
@@ -530,6 +530,30 @@ export default class IdsPopup extends Base {
...
+  get maxHeight(): string | null {
+    return this.getAttribute(attributes.MAX_HEIGHT);
+  }
+
+  /**
+   * Set the max height value
+   * @param {string | number | null} value The value
+   */
+  set maxHeight(value: string | number | null) {
+    let val: any = parseInt(value as string, 10);
```
Comment (danielortiz1982) [inline src/components/ids-popup/ids-popup.ts:545]: This looks good @deep7102, one question... Does `val` have to be `any`? Or can we possibly lock it down to the types that will be used/returned from the `parseInt()` fn? 
`let val: string | number  = parseInt(value as string, 10)` Just a suggestion to make better use of our types.

Comment (deep7102) [reply src/components/ids-popup/ids-popup.ts]: @danielortiz1982 Updated now to make better use of types.

**Review by tmcconechy — DISMISSED** [review]
Looks 🔥

**Review by tmcconechy — DISMISSED** [review]

**Review by tmcconechy — DISMISSED** [review]

**Review by jdilag90 — DISMISSED** [review]
Looks great nice work!

**Review by jdilag90 — APPROVED** [review]

**Review by tmcconechy — APPROVED** [review]

**Pull request merged**